cirq-core/cirq/optimizers/clifford_decomposition_test.py

```diff
@@ -0,0 +1,174 @@
+    expect_circ = cirq.Circuit(cirq.X(qubits[0]), cirq.H(qubits[0]), cirq.S(qubits[0]))
+    ops = cirq.decompose_clifford_tableau_to_operations(qubits, args.tableau)
+    circ = cirq.Circuit(ops)
+    assert_allclose_up_to_global_phase(cirq.unitary(expect_circ), cirq.unitary(circ), atol=1e-7)
+
+    qubits = cirq.LineQubit.range(1)
+    args = cirq.ActOnCliffordTableauArgs(
+        tableau=cirq.CliffordTableau(num_qubits=1),
+        axes=[0],
+        prng=np.random.RandomState(),
+        log_of_measurement_results={},
+    )
+    cirq.act_on(cirq.Z, args, allow_decompose=False)
+    cirq.act_on(cirq.H, args, allow_decompose=False)
+    cirq.act_on(cirq.S, args, allow_decompose=False)
+    cirq.act_on(cirq.H, args, allow_decompose=False)
+    cirq.act_on(cirq.X, args, allow_decompose=False)
+    expect_circ = cirq.Circuit(
+        cirq.Z(qubits[0]),
+        cirq.H(qubits[0]),
+        cirq.S(qubits[0]),
+        cirq.H(qubits[0]),
+        cirq.X(qubits[0]),
+    )
+    ops = cirq.decompose_clifford_tableau_to_operations(qubits, args.tableau)
+    circ = cirq.Circuit(ops)
+    assert_allclose_up_to_global_phase(cirq.unitary(expect_circ), cirq.unitary(circ), atol=1e-7)
+
+
+def test_clifford_decompose_two_qubits():
+    """Two random instance for two qubits decomposition."""
+    qubits = cirq.LineQubit.range(2)
+    args = cirq.ActOnCliffordTableauArgs(
+        tableau=cirq.CliffordTableau(num_qubits=2),
+        axes=[0],
+        prng=np.random.RandomState(),
+        log_of_measurement_results={},
+    )
+    cirq.act_on(cirq.H, args, allow_decompose=False)
+    args.axes = [0, 1]
+    cirq.act_on(cirq.CNOT, args, allow_decompose=False)
+    expect_circ = cirq.Circuit(cirq.H(qubits[0]), cirq.CNOT(qubits[0], qubits[1]))
+    ops = cirq.decompose_clifford_tableau_to_operations(qubits, args.tableau)
+    circ = cirq.Circuit(ops)
+    assert_allclose_up_to_global_phase(cirq.unitary(expect_circ), cirq.unitary(circ), atol=1e-7)
+
+    qubits = cirq.LineQubit.range(2)
+    args = cirq.ActOnCliffordTableauArgs(
+        tableau=cirq.CliffordTableau(num_qubits=2),
+        axes=[0],
+        prng=np.random.RandomState(),
+        log_of_measurement_results={},
+    )
+    cirq.act_on(cirq.H, args, allow_decompose=False)
+    args.axes = [0, 1]
+    cirq.act_on(cirq.CNOT, args, allow_decompose=False)
+    args.axes = [0]
+    cirq.act_on(cirq.H, args, allow_decompose=False)
+    cirq.act_on(cirq.S, args, allow_decompose=False)
+    args.axes = [1]
+    cirq.act_on(cirq.X, args, allow_decompose=False)
+    expect_circ = cirq.Circuit(
+        cirq.H(qubits[0]),
+        cirq.CNOT(qubits[0], qubits[1]),
+        cirq.H(qubits[0]),
+        cirq.S(qubits[0]),
+        cirq.X(qubits[1]),
+    )
+
+    ops = cirq.decompose_clifford_tableau_to_operations(qubits, args.tableau)
+    circ = cirq.Circuit(ops)
+    assert_allclose_up_to_global_phase(cirq.unitary(expect_circ), cirq.unitary(circ), atol=1e-7)
+
+
+def test_clifford_decompose_by_unitary():
```

Comment:
I'm wondering if we can reuse the implementation of `assert_implements_consistent_protocols` in `cirq.testing.consistent_protocols`? It does have a decomposition consistent with unitary assertion...

Reply:
After a brief look of this function, I think it is more suitable to put into the `n_qubit_clifford_gate` class, which is the next step when all these Clifford Tableau helper functions are done. Wdyt?

Reply:
I guess that can work - and will be consistent with having `Gate`s having a decomposition - but on the other hand there is nothing wrong with having it on the tableau and then the `Gate` delegating to the tableau's method, so I think I'm okay to have it here, as this way the Aaaronson-Gottesman paper implementations are kept closer together.